packages/block-editor/src/components/provider/test/use-block-sync.js

```diff
@@ -362,4 +362,78 @@ describe( 'useBlockSync hook', () => {
 		);
 		expect( onInput ).not.toHaveBeenCalled();
 	} );
+
+	it( 'should use fresh callbacks if onChange/onInput have been updated', async () => {
+		const fakeBlocks = [
+			{ clientId: 'a', innerBlocks: [], attributes: { foo: 1 } },
+		];
+		const onChange1 = jest.fn();
+		const onInput = jest.fn();
+
+		let registry;
+		const setRegistry = ( reg ) => {
+			registry = reg;
+		};
+		let root;
+		await act( async () => {
+			root = create(
+				<TestWrapper
+					setRegistry={ setRegistry }
+					value={ fakeBlocks }
+					onChange={ onChange1 }
+					onInput={ onInput }
+				/>
+			);
+		} );
+
+		// Create a persistent change.
+		registry
+			.dispatch( 'core/block-editor' )
+			.updateBlockAttributes( 'a', { foo: 2 } );
+
+		const updatedBlocks1 = [
+			{ clientId: 'a', innerBlocks: [], attributes: { foo: 2 } },
+		];
+
+		expect( onChange1 ).toHaveBeenCalledWith( updatedBlocks1, {
+			selectionEnd: {},
+			selectionStart: {},
+		} );
+
+		const newBlocks = [
+			{ clientId: 'b', innerBlocks: [], attributes: { foo: 1 } },
+		];
+
+		const onChange2 = jest.fn();
+
+		// Update the component to point at a "different entity" (e.g. different
+		// blocks and onChange handler.)
+		await act( async () => {
+			root.update(
+				<TestWrapper
+					setRegistry={ setRegistry }
+					value={ newBlocks }
+					onChange={ onChange2 }
+					onInput={ onInput }
+				/>
+			);
+		} );
+
+		// Create a persistent change.
+		registry
+			.dispatch( 'core/block-editor' )
+			.updateBlockAttributes( 'b', { foo: 3 } );
+
+		// The first callback should not be called with the new change.
+		expect( onChange1 ).toHaveBeenCalledWith( updatedBlocks1, {
```

Comment:
I don't think this checks that this was the only call to this function.

You might need to add `toHaveBeenCalledTimes` or other matchers.

Reply:
I don't totally understand why the first `onChange1` callback is firing at all here. the store change should not affect it. Perhaps there is still a stale callback hanging around? (in this case, it is ok because it is the same object as the previous change, so it's not receiving the new block control value)

are you saying it should be fine to check if it was just called once?

Reply:
Ah, my bad. I just needed to reset the mock before updating the component. Then we can check that it was not called at all.

Reply:
Yes